zilliqa/src/consensus.rs

```diff
@@ -435,6 +447,37 @@ impl Consensus {
             .map(|opt| opt.unwrap_or_default())
     }
 
+    pub fn get_txns_to_execute(&mut self) -> Vec<SignedTransaction> {
+        // Note: push back Txns that have an incorrect nonce
+        self.new_transactions
+            .values()
+            .cloned()
+            .filter(|tx| {
+                if self.state.has_account(tx.from_addr)
+                    && self.state.must_get_account(tx.from_addr).nonce != tx.transaction.nonce
```

Comment:
If the nonce is lower (or higher) it will re-try until max and then get dropped, yes. But that is correct - you don't want to execute Tx that isn't sequential.

There should be an additional check that nonces lower than the account nonce get rejected immediately, but that can't go in until this ticket is closed, now I think of it https://github.com/Zilliqa/zq2/issues/434

If you try to do it now, it's not quite right imo so it should be added in a follow up PR

Reply:
Fair enough, though I don't see why simply changing the `!=` to a `<` would be "not quite right", can you clarify what you mean here? E.g. if tx nonce is 0 but account nonce is 1, I don't see any reason in putting it into the retry loop.

If there's a good reason then I've definitely no issue with doing it in a followup, just trying to understand why the naive solution wouldn't work.

Reply:
Changing the `!=` to a `<` would be incorrect imo - this would return true for nonces arbitrarily higher than account nonce and so you could end up with TXs put into a block and then executed out of order (a block could contain txs with nonce 2 then 1 for example).

The 'not quite right' refers to never letting txs with nonces that are too low enter the mempool (which we should have), so it's kinda orthogonal comment.

In the tx not entering the mempool with nonce too low scenario IMO we should have accounts with a nonce 0 if you have sent funds to that address, that way you know whether the tx 'from' is a valid account (has funds).

Reply:
> this would return true for nonces arbitrarily higher than account nonce
But `!=` also returns true for tx nonces higher than account nonces... It returns true for anything which isn't equal to the account nonce. I think we might be talking about different things haha
So the way I understand it, this logic takes all the transactions inside `new_transaction` and, if their nonce doesn't exactly match the current account nonce, starts a retry counter for them in case they become valid later (and if they fail enough times for the counter to tick down to zero they're evicted from the mempool anyway). Right? But if an _account_ has a nonce already higher than a _transaction_, then the transaction will never become valid, so why even bother with the retry counter instead of just letting it fail immediately*?
I also don't see how this would lead to out-of-order transactions. If an account has nonce 1, and you submit txs with nonce 2 and 1, the first one would go into `new_transactions_waiting` (because account.nonce < tx.nonce) while the second one would get executed, incrementing the account's nonce. Next block, that first transaction will get executed (unless another tx with nonce 2 front-runs it, of course).

Actually, incidentally, this does prevent multiple transactions from getting executed for the same account in the same block, right? (Because all of them except the first one will have the "wrong" nonce.) I'm not sure if that's something we want, longer-term.

In any case let me know if any of this makes sense or if we've got a different understanding of things here.

*Which could be done with a shortcut check here, or even just letting it fallthrough to execution and getting rejected by the evm.

> In the tx not entering the mempool with nonce too low scenario IMO we should have accounts with a nonce 0 if you have sent funds to that address, that way you know whether the tx 'from' is a valid account (has funds).
I think this is orthogonal too, since nonces are initialised to 0 by default so this is just a shortcut to provide a fail-early I suppose. Not sure if that's a particularly useful shortcut - how often will transfers be sent from non-existent accounts, vs. from accounts who simply don't have enough funds, or even accounts who've transacted in the past but then had all their funds drained back to 0?

Reply:
We def have a different understanding ahah :)

I agree, txs with nonce too low should be immediately removed, I would just put it in another PR since it should also go with some logic to make sure accounts that have funds always exist (orthogonal topic). It should be that 1) They can not enter the mem pool at all with a nonce too low, or no funds at from address and 2) during this bit of code, if nonce is too low, it gets removed. If you like, I could put it in this PR

In your example of the 2 and 1 nonces - both of these return false when account.nonce is 1 for account.nonce < tx.nonce and so both would go into new_tx_waiting, right? That doesn't seem correct however I look at it since one is valid and another isn't.

Another unrelated todo: during these discussions I realised there is another edge case, there is no guard against multiple TXs from the same account but with the same nonce getting executed. I guess I'll do that tomorrow.

> I think this is orthogonal too, since nonces are initialised to 0 by default so this is just a shortcut to provide a fail-early I suppose. Not sure if that's a particularly useful shortcut - how often will transfers be sent from non-existent accounts, vs. from accounts who simply don't have enough funds, or even accounts who've transacted in the past but then had all their funds drained back to 0?

Oh, so you wouldn't fail these TXs (no from account funds)? IMO they are invalid, and invalid TXs are a way to spam the system. I do agree though that it's an edge case.

Reply:
>I agree, txs with nonce too low should be immediately removed, I would just put it in another PR since it should also go with some logic to make sure accounts that have funds always exist (orthogonal topic)

Fair enough, my only qualms would be that if one such transaction _is_ in the mempool, it'd start the retry counter for however many blocks instead of getting immediately evicted.
If we prevent such transactions from getting into the mempool in the first place, though, then sure.

> both of these return false when account.nonce is 1 for account.nonce < tx.nonce

No, I mean, `1 < 1` will return false and thus the transaction with nonce 1 will _not_ go into `new_transactions_waiting` but rather will _pass_ the filter and get executed. `1 < 2` will return true which will trigger the waiting logic for the tx with nonce 2.

Also to relate this to my earlier point, in this scenario a transaction with nonce 0 would then also go into the waiting pool with the current `!=`, even though it will never be valid henceforth. With `<` it will return false, thus be returned from `get_txns_to_execute()`, passed to the evm, fail and be evicted immediately. As I mentioned above if you just filter them out before they get into the mempool then fine - it just seems a bit odd to keep the logic that way, juuust in case there's ever a way for such txs to get into the mempool. Or to look at it another way: this waiting logic is specifically for giving special treatment to transactions whose nonces are higher than the current account nonce - so why not have the comparison check reflect that?

> Oh, so you wouldn't fail these TXs (no from account funds)? IMO they are invalid, and invalid TXs are a way to spam the system. I do agree though that it's an edge case.
On further thought I actually see where you're coming from - if the sending account has no funds, then it won't even be able to incur gas costs for sending the invalid transaction.
Ideally you'd fix that by shortcutting the balance check to make sure you can pay for gas, though: again an account could have a non-0 nonce but have all its funds drained, which would fail your check but still enable the spamming. Still, I think both checks might be useful.

> Another unrelated todo: during these discussions I realised there is another edge case, there is no guard against multiple TXs from the same account but with the same nonce getting executed. I guess I'll do that tomorrow.

Doesn't the account nonce get incremented after every transaction? As part of `apply_transaction()`. Then the next TX with the same nonce will fail the nonce check.
Unless you mean that you aren't rejecting them early here. In which case I agree, moreover a more sophisticated handling of the nonce check here would also let you potentially pass through several transactions for execution as long as they have contiguously increasing nonces and are in the correct order, rather than pushing them back to be retried next block. (A really good implementation would accept out-of-order transactions as long as they're contiguous and re-order them before returning the output vec of txs, but that's probably unnecessary until later.)

Reply:
Just chiming in to say that Ethereum clients will expect to be able to send multiple contiguous (by nonce) transactions in a row and expect them to be reordered and executed as you describe. Its something we can implement later, but we'll need to do it at some point before launch. I would guess they have limits on how long a transaction with too great a nonce will be kept in their mempool - I'll have a research.

Reply:
Given how complex this is starting to sound, I propose we merge this (because even this preliminary retry-counter, despite not handling many of the edge cases, is better than nothing) and then open an issue to hash out the exact details and implement it as a separate feature.

I'd still switch the comparison to `<` here (for the reasons mentioned above - it just fits the conceptual logic of what we're trying to do better) but it's not essential and if you still disagree I'm happy to merge as-is.

Reply:
> Fair enough, my only qualms would be that if one such transaction _is_ in the mempool, it'd start the retry counter for however many blocks instead of getting immediately evicted.

I am confused - I thought you previously said that txs with nonces too low should be immediately removed? 

> why even bother with the retry counter instead of just letting it fail immediately

Anyway sorry, yes, I messed up the 2 & 1 example just there by switching them accidentally. However, it seems you would suggest/are happy you allow txs with an incorrect nonce through to be executed by the next layer which seems incorrect to me(?)

However, it will not break the code (it checks the nonce is correct in exec.rs) and we are going to address in a follow up PR anyway, so I have changed it to `<` as you say :)

> Doesn't the account nonce get incremented after every transaction? As part of apply_transaction(). Then the next TX with the same nonce will fail the nonce check.

Yes, but the point is that you 'trick' the node into wasting resource, it is a form of spamming the system. Consider that there is a tx limit inside blocks. If you can submit 1 valid tx and 100 invalid ones (once the nonce increases), as we currently have it, you cause the node to waste its potential revenue stream (it could even be an attack by another validator to 'dilute' your mempool so it gets more of the profits). The system should be as efficient as possible imo and not do stuff like execute txns only to later find they're invalid (and no fee is received).

It seems the solution to all this is to have one of 3 things happen when collecting tx for execution. Either it gets selected to be executed, or it is invalid in some way (low nonce) and gets rejected, or it could be valid in the future and gets a timer put on it for staying too long.

Reply:
> However, it seems you would suggest/are happy you allow txs with an incorrect nonce through to be executed by the next layer which seems incorrect to me(?)

Not necessarily but I think the consensus is that the check to reject those should go into a separate PR, alongside work to e.g. reject transactions from empty accounts. All this is doing is kicking their rejection into that other PR, rather than letting them uselessly spin in the retry counter.

> Yes, but the point is that you 'trick' the node into wasting resource, it is a form of spamming the system.

This is true _iff_ the sending account has no funds. Otherwise, rejected transactions still charge gas, don't they? I assume they should: you charge for the EVM execution, even if the result of that execution was a reject reason rather than a state transition.

This is why I have no qualms with letting invalid transactions through, for now. We can add more filtering conditions as a UX improvement - _or_ to filter out accounts which currently do not have any funds to pay any fees even if rejected. Otherwise I don't think it's an issue.